pkg/process/metadata/workloadmeta/grpc.go

```diff
@@ -5,12 +5,154 @@
 
 package workloadmeta
 
-type mockableGrpcListener interface {
-	writeEvents(procsToDelete, procsToAdd []*ProcessEntity)
+import (
+	"net"
+	"strconv"
+	"sync"
+
+	"google.golang.org/grpc"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/proto/pbgo"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// GRPCServer implements a gRPC server to expose Process Entities collected with a WorkloadMetaExtractor
+type GRPCServer struct {
+	config    config.ConfigReader
+	extractor *WorkloadMetaExtractor
+	server    *grpc.Server
+	// The address of the server set by start(). Primarily used for testing. May be nil if start() has not been called.
+	addr net.Addr
+
+	wg sync.WaitGroup
+}
+
+// NewGRPCServer creates a new instance of a GRPCServer
+func NewGRPCServer(config config.ConfigReader, extractor *WorkloadMetaExtractor) *GRPCServer {
+	l := &GRPCServer{
+		config:    config,
+		extractor: extractor,
+		server:    grpc.NewServer(),
+	}
+
+	pbgo.RegisterProcessEntityStreamServer(l.server, l)
+	return l
+}
+
+func (l *GRPCServer) consumeProcessDiff(diff *ProcessCacheDiff) ([]*pbgo.ProcessEventSet, []*pbgo.ProcessEventUnset) {
+	setEvents := make([]*pbgo.ProcessEventSet, len(diff.creation))
+	for i, proc := range diff.creation {
+		setEvents[i] = &pbgo.ProcessEventSet{
+			Pid:      proc.pid,
+			Language: &pbgo.Language{Name: string(proc.language.Name)},
+		}
+	}
+
+	unsetEvents := make([]*pbgo.ProcessEventUnset, len(diff.deletion))
+	for i, proc := range diff.deletion {
+		unsetEvents[i] = &pbgo.ProcessEventUnset{Pid: proc.pid}
+	}
+
+	return setEvents, unsetEvents
+}
+
+// Start starts the GRPCServer to listen for new connections
+func (l *GRPCServer) Start() error {
```

Comment:
Do we need TLS in the future ? We used it for remote workloadmeta

Reply:
We probably want to add TLS to the server indeed. We're tracking this work separately in a future PR.